### ISSUE TITLE
Allow a connect timeout to be specified

### ISSUE THREAD

**Issue by mattiash** [axios/axios#1739]
#### Summary

I want to use axios to connect to an http-server that is available via several different servers with different url:s. If one of the servers does not respond, I want to try another server instead. To get fast failovers, I want to set a low timeout for the request.

But the server sometimes takes a long time to respond (when I am making a complex query). To give the server time to generate the response, I have to set a high timeout.

What I really want to do is set a low connectTimeout, so that if the server has not accepted the tcp connection within this timeout, the request is aborted, but a high timeout to give the server time to generate the response.

#### Context

- axios version: *v0.18.0*
- Environment: *node v8.11.3*

PR coming up.

**Comment by SteffenLanger** [comment]
Hi @matthiash,

The idea is great. What is the current status of this issue and can I help you or the axios team somehow?

**Comment by mattiash** [comment]
Well, I haven't gotten any feedback on my PR from anyone on the axios team. I solved my problem by using got instead. It allows me to specify timeouts for connect and request separately: https://github.com/sindresorhus/got#timeout

**Comment by benb1983** [comment]
+1 this is big problem

**Comment by gwuah** [comment]
@benb1983 Can you give an example of how to implement that with `CancelToken` ?

**Comment by benb1983** [comment]
@gwuah something like this...

`
let response = null
const source = axios.CancelToken.source()

        setTimeout(() => {
          if (response === null) {
            source.cancel()
          }
        }, 5000) // connection timeout here in ms (5 seconds)

response = await axios.post(this.authUrl, this.authData, {
        ...headers,
       cancelToken: source.token
      })


`

**Comment by cb-manideep** [comment]
+1

**Comment by lixaotec** [comment]
+1

**Comment by jasonsaayman** [comment]
Hi,

I looked at the pull request, I think this is not the way to go about making this change. I will have a look in the future at how we can solve this. For now we are working on a release, after which we will look at this again.

Thanks

**Comment by dstokesf5** [comment]
@jasonsaayman, any updates on when this issue might be revisited?

**Comment by tdaniely** [comment]
In node you can do this by configuring the agents in the defaults, in an instance, or per request.
In the browser, I don't think a web page has any control, it's completely the browsers decision.

```javascript
import axios from "axios";
import http from "http";
import https from "https";

axios.defaults.httpAgent = new http.Agent({ timeout: 5000 });
axios.defaults.httpsAgent = new https.Agent({ timeout: 5000 });

let client = axios.create({
    httpAgent: new http.Agent({ timeout: 5000 }),
    httpsAgent: new https.Agent({ timeout: 5000 }),
});

let response = axios.request({
    httpAgent: new http.Agent({ timeout: 5000 }),
    httpsAgent: new https.Agent({ timeout: 5000 }),
});
```

**Comment by Spasfonx** [comment]
Thanks @tdaniely this workaround with the `Agent` works perfectly 👌  we can differanciate connect timeout and request timeout. The only drawback is we have a wrong error message, the value in the error message is the `timeout` specified in axios config value instead of the correct `Agent` timeout value : 

```javascript
axios.defaults.timeout = 3000
axios.defaults.httpAgent = new http.Agent({ timeout: 500 })
axios.defaults.httpsAgent = new https.Agent({ timeout: 500 })

axios.get("https://www.google.com:81")
   .catch(console.log) // display 'timeout of 3000ms exceeded' but timeout after 500ms
```

**Comment by tdaniely** [comment]
Yes I was checking the timeout behaviour and I noticed some weird things in the code.
Specifically most of it is in not even axios but [follow-redirects](https://github.com/follow-redirects/follow-redirects).
And it's throwing fake os error ECONNABORTED when the failure is applicative, which is even weirder.
Like you I would expect to get a specific error for a response timeout and a specific error for the tcp handshake timeout.
And this is especially important to detect if we are blocked by DDOS protection, or the server actually didn't respond.
Might want to open a seperate issue about the timeout throw handling.
I'm not contributing to axios, this is just what I see.

**Comment by jonmcnamaradale** [comment]
Thank you @tdaniely this helped me too, I couldn't understand why I wasn't getting a timeout but it is working fine now with your tip.

**Comment by huytran1ibm** [comment]
same here, we are experiencing ECONNABORTED errors

**Comment by borovik96** [comment]
Does anybody have reproducing code in sandbox with "no timeout" behaviour? I have production code where it happens some time, but can not reproduce it by myself

**Comment by serg06** [comment]
Before Axios V1 you can do:

```ts
axios.get("https://www.google.com:81", {
    httpAgent: http.Agent({ timeout: 500 }),
    httpsAgent: https.Agent({ timeout: 500 }),
});
```

I'm not sure if this works with Axios V1.
